packages/store/src/database/backends/indexedDB.ts

```diff
@@ -1,48 +1,322 @@
+  ) {
+    await this.db.transaction(
+      "readwrite",
+      this.db.entities,
+      this.db.derived_taskComponents,
+      async () => {
+        const rows = await this.db.entities.where("id").anyOf(ids).toArray();
+
+        const rowMapping = rows.reduce((map, row) => {
+          map.set(row.id, row as DexieEntityRowWithPrimaryKey);
+          return map;
+        }, new Map<EntityID, DexieEntityRowWithPrimaryKey>());
+
+        const transformedEntities = transformer(rowMapping).values();
+        const transformedRows = Array<DexieEntityRowWithPrimaryKey>();
+        for (const row of transformedEntities) {
+          transformedRows.push({
+            rowID: row.rowID,
+            id: row.id,
+            lastEventID: row.lastEventID,
+            data: row.data,
+          });
+        }
+        await this.db.entities.bulkPut(transformedRows);
+      },
+    );
+  }
+
+  async putEntities(
+    entities: DatabaseBackendEntityRecord<Entity>[],
+  ): Promise<void> {
+    await this.db.transaction(
+      "readwrite",
+      this.db.entities,
+      this.db.derived_taskComponents,
+      async () => {
+        const newEntities = entities.map((record) => ({
+          id: record.entity.id,
+          lastEventID: record.lastEventID,
+          data: JSON.stringify(record.entity),
+        }));
+
+        await this.db.entities.bulkPut(newEntities);
+      },
+    );
   }
 
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  putEntities(entities: DatabaseBackendEntityRecord<Entity>[]): Promise<void> {
-    return Promise.resolve(undefined);
+  async putEvents(events: Event[]): Promise<void> {
+    this.db.transaction("rw", this.db.events, async () => {
+      await this.db.events.bulkAdd(
+        events.map((event) => ({
+          entityID: event.entityID,
+          id: event.id,
+          data: JSON.stringify(event),
+        })),
+      );
+    });
   }
 
-  // @ts-ignore
-  // eslint-disable-next-line @typescript-eslint/no-unused-vars
-  putEvents(events: Event[]): Promise<void> {
-    return Promise.resolve(undefined);
+  async _fetchPrimaryKeyFromUniqueKey<Row, PK>(
+    table: Dexie.Table<Row, PK>,
+    key: string,
+    value: IndexableType,
+  ): Promise<PK> {
+    // there will never be more then one key
+    const afterRowPrimaryKeys = await table
+      .where(key)
+      .equals(value)
+      .primaryKeys();
+    return afterRowPrimaryKeys[0];
   }
 }
+
+function compareUsingPredicate<
+  Row,
+  PK,
+  Key extends string,
+  Value extends IndexableType,
+>(clause: WhereClause<Row, PK>, predicate: DatabaseQueryPredicate<Key, Value>) {
+  switch (predicate[1]) {
+    case "<":
+      return clause.below(predicate[2]);
+    case "<=":
+      return clause.belowOrEqual(predicate[2]);
+    case "=":
+      return clause.equals(predicate[2]);
+    case ">":
+      return clause.above(predicate[2]);
+    case ">=":
+      return clause.aboveOrEqual(predicate[2]);
+  }
+}
```

Comment:
Not too keen on this helper function. We could use `comparseUsingPredicate()` function instead, but my understanding is that we will no longer leverage the indexes in those cases...

Reply:
Just feels silly to map between the two, but I'll get over it especially since we can remove `compareUsingPredicate` now with your new optimized query.

Reply:
Not that it's relevant now, but you might prefer:

```
const fn = {
["<"]: clause.below,
["<=]: clause.belowOrEqual,
// ...
}[predicate[1]];
return fn.bind(clause)(predicate[2]);
```